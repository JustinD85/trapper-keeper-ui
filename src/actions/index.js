export const ADD_NEW_NOTE = "ADD_NEW_NOTE";
export const addNewNote = newNote => ({
  type: ADD_NEW_NOTE,
  newNote
});

export const ADD_ALL_NOTES = "ADD_ALL_NOTES";
export const addAllNotes = notes => ({
  type: ADD_ALL_NOTES,
  notes
});

export const SET_LOADING = "SET_LOADING";
export const setLoading = isLoading => ({
  type: SET_LOADING,
  isLoading
});

export const SET_ERROR = "SET_ERROR";
export const setError = error => ({
  type: SET_ERROR,
  error
});