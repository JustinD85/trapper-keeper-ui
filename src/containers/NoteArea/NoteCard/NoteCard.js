import React from "react";
import { NoteItems } from "../../";
class NoteCard extends React.Component {
  constructor({ note }) {
    super(note);
    this.state = {
      title: note.title,
      noteItems: note.noteItems
    };
  }

  render() {
    const { title, noteItems } = this.props.note;
    return (
      <div>
        <h2>{title}</h2>
        <ul>
          <NoteItems noteItems={noteItems} />
        </ul>
      </div>
    );
  }
}

export default NoteCard;
