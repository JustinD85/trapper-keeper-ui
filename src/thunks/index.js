import {
  addAllNotes,
  setLoading,
  setError,
  addNewNote,
  updateNote,
  removeNote
} from "../actions";
import { fetchCall } from "../utils/api";

const baseUrl = "http://localhost:3001/notes";

export const getOptions = (method, data) => ({
  method,
  headers: {
    "Content-Type": "application/json"
  },
  body: JSON.stringify(data)
});

export const getNotes = () => {
  return async dispatch => {
    try {
      dispatch(setLoading(true));
      const notes = await fetchCall(baseUrl);
      dispatch(setLoading(false));
      dispatch(addAllNotes(notes));
    } catch (error) {
      dispatch(setError("Error fetching notes"));
    }
  };
};

export const postNote = newNote => {
  return async dispatch => {
    try {
      dispatch(setLoading(true));
      let note = await fetchCall(baseUrl, getOptions("POST", newNote));
      dispatch(setLoading(false));
      dispatch(addNewNote(note));
    } catch (error) {
      dispatch(setError(error.message));
    }
  };
};

export const putNote = updatedNote => {
  return async dispatch => {
    try {
      dispatch(setLoading(true));
      const note = await fetchCall(
        `${baseUrl}/${updatedNote.id}`,
        getOptions("PUT", updatedNote)
      );
      dispatch(setLoading(false));
      dispatch(updateNote(note));
    } catch (error) {
      dispatch(setError(error.message));
    }
  };
};

export const deleteNote = newNote => {
  return async dispatch => {
    try {
      dispatch(setLoading(true));
      await fetchCall(`${baseUrl}/${newNote.id}`, { method: 'DELETE' });
      dispatch(setLoading(false));
      dispatch(removeNote(newNote.id));
    } catch (error) {
      dispatch(setError(error.message));
    }
  };
};
