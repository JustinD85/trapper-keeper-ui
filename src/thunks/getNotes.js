import  { addNotes }  from '../actions';
import { fetchCall } from '../utils/api';

export const getNotes = (url) => {
  return async (dispatch) => {
    const data = await fetchCall(url)
    console.log('from thunk', data)
    dispatch(addNotes(data))   
  }
}