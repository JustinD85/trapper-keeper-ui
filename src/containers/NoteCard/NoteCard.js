import React from "react";
import { NoteItems } from "../";
import PropTypes from "prop-types";
import { Link, NavLink } from "react-router-dom";
import { deleteNote } from '../../thunks';
import { connect } from 'react-redux';
import { Card, IconButton, CardHeader, CardContent, Tooltip } from '@material-ui/core';
import { DeleteOutline } from '@material-ui/icons'
import { withStyles } from '@material-ui/core/styles'

const styles = {
  link: {
    textDecoration: 'none',
  },
  card: {
    margin: 10,
  },
  title: {
    color: '#48494a',
  },
  delete: {
    float: 'right',
    margin: '3px'
  }
}

export const NoteCard = ({ note, deleteNote, classes }) => (
  <Tooltip title='Click to edit!' placement='top' enterDelay={2000}>
    <Card className={classes.card}>
      <Link to='/' className={classes.delete}>
        <Tooltip title='Delete this note!'>
          <IconButton onClick={() => deleteNote(note)}>
            <DeleteOutline />
          </IconButton>
        </Tooltip>
      </Link>
      <Link to={`/notes/${note.id}`} className={classes.link}>
        <CardHeader
          title={note.title}
          classes={{ title: classes.title }}
        />
        <CardContent>
          <ul>
            <NoteItems noteItems={note.noteItems} />
          </ul>
        </CardContent>
      </Link>
    </Card>
  </Tooltip>
);

export const mapDispatchToProps = (dispatch) => ({
  deleteNote: (note) => dispatch(deleteNote(note))
});

export default withStyles(styles)(connect(null, mapDispatchToProps)(NoteCard));

NoteCard.propTypes = {
  note: PropTypes.object,
  deleteNote: PropTypes.func
};
