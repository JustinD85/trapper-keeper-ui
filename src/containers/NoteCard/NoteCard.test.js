import React from "react";
import { shallow} from "enzyme";
import { NoteCard, mapDispatchToProps } from "./NoteCard";
import { deleteNote } from "../../thunks";
import { removeNote } from "../../actions";
jest.mock("../../thunks");
deleteNote.mockReturnValue(removeNote(1));

describe("NoteCard", () => {
  let wrapper;
  let deleteNoteMock = jest.fn();
  const mockNote = {
    id: 1,
    title: "Best Note",
    noteItems: [{ text: "lala" }, { text: "hooray" }]
  };

  beforeEach(() => {
    wrapper = shallow(<NoteCard note={mockNote} classes={{card:''}} deleteNote={deleteNoteMock} />)
  });

  it("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should call delete note when clicked", () => {
    wrapper = shallow(<NoteCard note={mockNote} classes={{card:''}} deleteNote={deleteNoteMock} />)
    wrapper.find("WithStyles(IconButton)").simulate("click");
    expect(deleteNoteMock).toHaveBeenCalled();
  });

  describe("mapDispatchToProps", () => {
    const note = { text: "note", id: 1 };
    const mockDispatch = jest.fn()
    const props = mapDispatchToProps(mockDispatch)
    props.deleteNote(note)
    expect(deleteNote).toHaveBeenCalled()
  });
});
