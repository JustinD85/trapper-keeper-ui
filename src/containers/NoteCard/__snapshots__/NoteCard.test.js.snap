// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteCard should match the snapshot 1`] = `
<WithStyles(Tooltip)
  enterDelay={2000}
  placement="top"
  title="Click to edit!"
>
  <WithStyles(Card)
    className=""
  >
    <Link
      replace={false}
      to="/"
    >
      <WithStyles(Tooltip)
        title="Delete this note!"
      >
        <WithStyles(IconButton)
          onClick={[Function]}
        >
          <pure(DeleteOutlineIcon) />
        </WithStyles(IconButton)>
      </WithStyles(Tooltip)>
    </Link>
    <Link
      replace={false}
      to="/notes/1"
    >
      <WithStyles(CardHeader)
        classes={
          Object {
            "title": undefined,
          }
        }
        title="Best Note"
      />
      <WithStyles(CardContent)>
        <ul>
          <WithStyles(NoteItems)
            noteItems={
              Array [
                Object {
                  "text": "lala",
                },
                Object {
                  "text": "hooray",
                },
              ]
            }
          />
        </ul>
      </WithStyles(CardContent)>
    </Link>
  </WithStyles(Card)>
</WithStyles(Tooltip)>
`;
